### PR TITLE
DAOS-10188 dtx: handle race between DTX aggregation and DTX refresh

### PR DESCRIPTION
**Pull request by Nasf-Fan**
There is race between DTX aggregation and DTX refresh. Consider the
following scenario:

The DTX leader triggers DTX commit for some DTX entry, then related
DTX participants (including the leader itself) will commit the DTX
entry on each own target in parallel. It is possible that the leader
has already committed locally but DTX aggregation removed the committed
DTX very shortly after the commit. On the other hand, on some non-leader
before the local commit, someone triggers DTX refresh for such DTX on
such non-leader. Unfortunately the DTX entry has already gone on the
leader. Then the non-leader will get -DER_TX_UNCERTAIN, that will cause
related applicaiton to fail unexpectedly.

So even if the system has DRAM pressue, we still need to keep some very
recent committed DTX entries to handle above race.

Signed-off-by: Fan Yong <fan.yong@intel.com>

### CODE REVIEW

File: src/vos/vos_layout.h [hunk 1]
```diff
@@ -147,8 +147,6 @@ struct vos_dtx_cmt_ent_df {
 	struct dtx_id			dce_xid;
 	/** The epoch# for the DTX. */
 	daos_epoch_t			dce_epoch;
-	/** The time of the DTX being handled on the server. */
-	daos_epoch_t			dce_handle_time;
```
Comment (jolivier23) [inline src/vos/vos_layout.h:151]: Anyway to do this while maintaining compatibility?   We are not supposed to break durable format between 2.0 and 2.2

**Review by jolivier23 — CHANGES_REQUESTED** [review]

File: src/vos/vos_layout.h (hunk 1 above)
Comment (Nasf-Fan) [reply src/vos/vos_layout.h:151]: So we have to keep these useless fields until 2.4 or 2.6?

Comment (jolivier23) [reply src/vos/vos_layout.h:151]: Ok, I just talked to Johann.

"The path is 1. Upgrade to 2.2 and still have the bug. 2. Dmg pool upgrade and the problem is fixed (but no way back)."

So, the code will need to check the pool version and not fail on the old one (as it does now).   Then, if the new version is there, we have the bug fixed.

**Review by jolivier23 — APPROVED** [review]

File: src/dtx/dtx_internal.h [hunk 2]
```diff
@@ -94,6 +94,21 @@ extern uint32_t dtx_agg_thd_cnt_lo;
...
+ * It is possible that the leader has already committed locally but DTX aggregation removed
+ * the committed DTX very shortly after the commit. On the other hand, on some non-leader
+ * before the local commit, someone triggers DTX refresh for such DTX on such non-leader.
+ * Unfortunately the DTX entry has already gone on the leader. Then the non-leader will
+ * get -DER_TX_UNCERTAIN, that will cause related application to fail unexpectedly.
+ *
+ * So even if the system has DRAM pressure, we still need to keep some very recent committed
+ * DTX entries to handle above race.
+ */
+#define DTX_AGG_AGE_PRESERVE	3
```
Comment (NiuYawei) [inline src/dtx/dtx_internal.h:110]: Even with this extra delay, refresh still could happen after the entry being aggregated on leader, right? How will we handle this? Is this a temporary workaround or final solution? What about aggregating only globally committed entry on leader?

**Review by NiuYawei — APPROVED** [review]

File: src/dtx/dtx_internal.h (hunk 2 above)
Comment (Nasf-Fan) [reply src/dtx/dtx_internal.h:110]: This extra delay is for the case of DTX commit difference on related targets, usually, as long as no target corruption, 3 seconds is enough for DTX committed on all related targets. After the global DTX commit, there will be no DTX refresh unless some target missed the DTX commit RPC. Under such case, if the committed DTX entry is removed by DTX aggregation, then the UNCERTAIN issue will happen again.

Make the DTX aggregation globally can resolve the issue, but the cost  may be too much, it needs additional RPCs. So we do not make such decision yet.

**Pull request merged**